src/lpython/semantics/python_ast_to_asr.cpp

```diff
@@ -3452,6 +3452,72 @@ class BodyVisitor : public CommonVisitor<BodyVisitor> {
         return nullptr;
     }
 
+    ASR::asr_t* handle_intrinsic_bool(Allocator &al, Vec<ASR::call_arg_t> args,
+                                        const Location &loc) {
+        ASR::expr_t *arg = nullptr, *value = nullptr;
+        ASR::ttype_t *type = nullptr;
+        if (args.size() > 0) {
+            arg = args[0].m_value;
+            type = ASRUtils::expr_type(arg);
+        }
+        ASR::ttype_t *to_type = ASRUtils::TYPE(ASR::make_Logical_t(al, loc,
+                                    4, nullptr, 0));
+        if (!arg) {
+            return ASR::make_LogicalConstant_t(al, loc, false, to_type);
+        }
+        if (ASRUtils::is_integer(*type)) {
+            if (ASRUtils::expr_value(arg) != nullptr) {
+                bool b = ASR::down_cast<ASR::IntegerConstant_t>(
+                                        ASRUtils::expr_value(arg))->m_n;
+                value = ASR::down_cast<ASR::expr_t>(make_LogicalConstant_t(al,
+                                loc, b, to_type));
+            }
+            return (ASR::asr_t *)ASR::down_cast<ASR::expr_t>(ASR::make_Cast_t(
+                al, loc, arg, ASR::cast_kindType::IntegerToLogical, to_type, value));
+
+        } else if (ASRUtils::is_real(*type)) {
+            if (ASRUtils::expr_value(arg) != nullptr) {
+                bool b = ASR::down_cast<ASR::RealConstant_t>(
+                                        ASRUtils::expr_value(arg))->m_r;
+                value = ASR::down_cast<ASR::expr_t>(make_LogicalConstant_t(al,
+                                loc, b, to_type));
+            }
+            ASR::ttype_t *int_type = ASRUtils::TYPE(ASR::make_Integer_t(al, loc,
+                                        4, nullptr, 0));
+            ASR::expr_t *t = ASR::down_cast<ASR::expr_t>(ASR::make_Cast_t(
+                al, loc, arg, ASR::cast_kindType::RealToInteger, int_type, nullptr));
+
+            return (ASR::asr_t *)ASR::down_cast<ASR::expr_t>(ASR::make_Cast_t(
+                al, loc, t, ASR::cast_kindType::IntegerToLogical, to_type, value));
+
+        } else if (ASRUtils::is_character(*type)) {
+            if (ASRUtils::expr_value(arg) != nullptr) {
+                char *c = ASR::down_cast<ASR::StringConstant_t>(
+                                        ASRUtils::expr_value(arg))->m_s;
+                value = ASR::down_cast<ASR::expr_t>(make_LogicalConstant_t(al,
+                                loc, std::string(c) != "", to_type));
+            }
+            // TODO: make bool() work for non-constant strings
```

Comment:
How do you wanna do this? Call a runtime function or have a node for this, too?

Reply:
Can this be done with the existing `Cast` ASR node? Such as casting Integer to Bool.

Reply:
I would suggest using `Cast` node. What do you all think?

Reply:
How about we add `CharacterToLogical` and `ComplexToLogical` under `cast_kind`? 

> Can this be done with the existing `Cast` ASR node?

I don't see any existing node that does the job here.

Reply:
Well here’s what I think.
1. First check whether the call is being made to `bool`. In an abstract sense, I mean verify that the semantics of the code is to call `bool` over a single input.
2. Now, see if the input is of intrinsic type (i.e., integer, real/float, complex). If so then things are pretty much clear, source type is the type of input argument and destination type is logical. If the input type is say a class then I think Python provides some magic method (I think it is `__bool__`) for conversion. We will have to call `__bool__`.
3. On the basis of step 2 either create a `Cast` node (if the type is node) or a `FunctionCall` (if the type is class).

I think I have covered all the cases for `bool(x)` expression. It is fairly easy to implement the above steps in ASR (provided I am not missing out on anything).

Reply:
> How about we add CharacterToLogical and ComplexToLogical under cast_kind?

I think yes we should add and use them in step 3 of my comment above.

Reply:
For non-constant strings in the LLVM backend we might not need a new C runtime function. We can call already existing function which computes the length of the string. The integer received in return will easily determine the result. If 0 then false and true in all other cases.